[AOTI] Fix #140546 and support AOTI package load for Intel GPU.

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #140686
* __->__ #140664
* #140269
* #140268
* #135320
* #135318
* #139026 


Fix #140546



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

EikanWang commented on torch/csrc/inductor/aoti_runner/model_container_runner_xpu.cpp:
```diff
@@ -17,18 +17,30 @@ AOTIModelContainerRunnerXpu::AOTIModelContainerRunnerXpu(
+    const std::string& model_so_path,
+    size_t num_models,
+    const std::string& device_str,
+    const std::string& cubin_dir) {
```
```suggestion
    const std::string& kernel_bin_dir) {
```

etaf replied:
resolved

EikanWang commented on torch/csrc/inductor/aoti_runner/model_container_runner.h:
```diff
@@ -22,11 +22,13 @@ class TORCH_API AOTIModelContainerRunner {
+  virtual ~AOTIModelContainerRunner();
 
-  std::vector<at::Tensor> run(
+  virtual std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs);
```
It changes the API definition. According to the issue description, the `null` stream means the current stream for CUDA. The same semantic should also be applicable for Intel GPU. Since this PR has fixed the virtual function routine, the runner instance could dispatch `run` function to the right device container runner.

etaf replied:
This relates to our definition of the run API. At first I thought it would be more generic to define it as `std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs, AOTInductorStreamHandle stream_handle = nullptr);`
. But this means that in `model_container_runner_xpu.h`, `model_container_runner_xpu.h`, `model_container_runner_cpu.h`,  we need to define this override API as well. Since `AOTIModelContainerRunnerCpu/Xpu/Cuda` will be used in pybind,  we'll get compilation error: 
```
torch/csrc/inductor/aoti_runner/pybind.cpp:21:11: required from here
/usr/include/c++/11/type_traits:1422:38: error: invalid use of incomplete type ‘struct AOTInductorStreamOpaque’
1422 | : public integral_constant<bool, __is_base_of(_Base, _Derived)>
| ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/xinanlin/xinanlin/pytorch/torch/csrc/inductor/aoti_runner/model_container_runner.h:5,
from torch/csrc/inductor/aoti_runner/model_container_runner_cpu.h:4,
from torch/csrc/inductor/aoti_runner/pybind.cpp:1:
torch/csrc/inductor/aoti_runtime/interface.h:16:8: note: forward declaration of ‘struct AOTInductorStreamOpaque’
```
The current design of this PR defines the interface of cpu/xpu/cuda container runner as `std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs);` is just keep the same before PR, so I think there is no issue to land it.
 
And if we prefer the interface `run` as `std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs, AOTInductorStreamHandle stream_handle = nullptr);` ,  we can create an adapter between pybind interface and the AOTI runner container for the container to resolve the above compilation issue in a new PR.

desertfire replied:
Agree with @EikanWang , this is an API breaking change. I think if you declare `stream_handle` as `void*`, it should solve your compilation issue.

etaf replied:
Hi, @desertfire I've updated the PR, could you please have a review? BTW could you please also review this series of PRs? They are all for AOTi XPU and are ready for review.

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf commented on torch/_export/__init__.py:
```diff
@@ -366,7 +366,7 @@ def optimized(*args, **kwargs):
         flat_inputs = pytree.tree_flatten((args, reorder_kwargs(kwargs, in_spec)))[0]
         flat_inputs = [x for x in flat_inputs if isinstance(x, torch.Tensor)]
-        flat_outputs = runner.run(flat_inputs)  # type: ignore[attr-defined]
+        flat_outputs = runner.run(flat_inputs, None)  # type: ignore[attr-defined]
```
Since we've changed the runner API from `std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs)` to `std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs, void* stream_handle = nullptr)`, Here wiil be error like:
```
ERROR: test_amp_fallback_random_cpu (__main__.AOTInductorTestABICompatibleCpu)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/xinanlin/xinanlin/pytorch/torch/testing/_internal/common_utils.py", line 3062, in wrapper
    method(*args, **kwargs)
  File "/home/xinanlin/xinanlin/pytorch/test/inductor/test_torchinductor.py", line 11757, in new_test
    return value(self)
  File "/home/xinanlin/xinanlin/pytorch/test/inductor/test_aot_inductor.py", line 524, in test_amp_fallback_random
    self.check_model(fn, example_inputs)
  File "/home/xinanlin/xinanlin/pytorch/test/inductor/test_aot_inductor.py", line 138, in check_model
    actual = AOTIRunnerUtil.run(
  File "/home/xinanlin/xinanlin/pytorch/test/inductor/test_aot_inductor_utils.py", line 132, in run
    return optimized(*example_inputs)
  File "/home/xinanlin/xinanlin/pytorch/torch/_export/__init__.py", line 369, in optimized
    flat_outputs = runner.run(flat_inputs)  # type: ignore[attr-defined]
TypeError: run(): incompatible function arguments. The following argument types are supported:
    1. (self: torch._C._aoti.AOTIModelContainerRunnerCpu, arg0: list[torch.Tensor], arg1: capsule) -> list[torch.Tensor]

Invoked with: <torch._C._aoti.AOTIModelContainerRunnerCpu object at 0x7fc168a5fdf0>, [tensor([[-0.1117, -0.4966,  0.1631, -0.8817,  0.0539,  0.6684, -0.0597, -0.4675
,
```
The Pybind Python API `run` does not know how to process the default void* stream_handle = nullptr, So we explicitly assign a None for it.

etaf replied:
@desertfire  @EikanWang , After changed the API as you suggested, we also get the following error, which use `runner.run(input)` but the changed Cpp API `std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs, void* stream_handle = nullptr)` can not handle this. The Pybind Python API run does not know how to process the default void* stream_handle = nullptr, So we need to explicitly assign a None for it. This is a real API break change now.
I think my previous design maybe better?
======================================================================
ERROR: test_non_tensor_input_cpu_with_stack_allocation (__main__.AOTInductorTestABICompatibleCpuWithStackAllocation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/xinanlin/xinanlin/pytorch/torch/testing/_internal/common_utils.py", line 3095, in wrapper
    method(*args, **kwargs)
  File "/home/xinanlin/xinanlin/pytorch/test/inductor/test_torchinductor.py", line 11822, in new_test
    return value(self)
  File "/home/xinanlin/xinanlin/pytorch/test/inductor/test_aot_inductor.py", line 1570, in test_non_tensor_input
    res = kernel_runner.run([a, b])
TypeError: run(): incompatible function arguments. The following argument types are supported:
    1. (self: torch._C._aoti.AOTIModelContainerRunnerCpu, arg0: list[torch.Tensor], arg1: capsule) -> list[torch.Tensor]

Invoked with: <torch._C._aoti.AOTIModelContainerRunnerCpu object at 0x7fcf7182ce30>, [tensor([ 0.0461,  0.4024, -1.0115,  0.2167, -0.6123,  0.5036,  0.2310,  0.6931,
        -0.2669,  2.1785]), tensor([ 0.1021, -0.2590, -0.1549, -1.3706, -0.1319,  0.8848, -0.2611,  0.6104,
        -0.0098, -1.4473])]

To execute this test, run the following from the base repo dir:
    python test/inductor/test_aot_inductor_arrayref.py AOTInductorTestABICompatibleCpuWithStackAllocation.test_non_tensor_input_cpu_with_stack_allocation

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

desertfire replied:
The cpp-side AOTI runner API should be the main integration point for AOTI, since after all it is about deploying models in cpp. I consider these pybind interfaces as internal APIs, so I am fine with breaking them instead of breaking the main runner API in cpp. 

The reason why we had these pybind interfaces in the first place is to facilitate Python side runner interface, previously `torch._export.aot_load` and more recently `torch._inductor.aoti_load_package`. You may need to update their implementation because of the issue you are mentioning here.

etaf replied:
> The cpp-side AOTI runner API should be the main integration point for AOTI, since after all it is about deploying models in cpp. I consider these pybind interfaces as internal APIs, so I am fine with breaking them instead of breaking the main runner API in cpp.
> 
> The reason why we had these pybind interfaces in the first place is to facilitate Python side runner interface, previously `torch._export.aot_load` and more recently `torch._inductor.aoti_load_package`. You may need to update their implementation because of the issue you are mentioning here.

@desertfire thanks for your time, I've updated the interface to support `void* stream_handle=nullptr`.

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_package/model_package_loader.h:
```diff
@@ -14,7 +14,9 @@ class TORCH_API AOTIModelPackageLoader {
-  std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs);
+  std::vector<at::Tensor> run(
+      const std::vector<at::Tensor>& inputs,
+      void* stream_handl = nullptr);
```
Nit: stream_handl-> stream_handle

etaf replied:
Resolved, thanks~

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

desertfire commented:
Hi @etaf, how much longer do you think you will need to land this one? Because some of the changes here are not XPU specific, and I want to make sure they can make the 2.6 branch cut (12/9).

etaf commented:
> Hi @etaf, how much longer do you think you will need to land this one? Because some of the changes here are not XPU specific, and I want to make sure they can make the 2.6 branch cut (12/9).

Hi, @desertfire, I'll try to land it today.

etaf commented on torch/csrc/inductor/aoti_runner/model_container_runner_cpu.h:
```diff
@@ -10,9 +10,11 @@ class TORCH_API AOTIModelContainerRunnerCpu : public AOTIModelContainerRunner {
+  ~AOTIModelContainerRunnerCpu() override;
 
-  std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs);
+  std::vector<at::Tensor> run(
```
Sorry, @desertfire, I don't get your point, here `std::vector<at::Tensor> run(const std::vector<at::Tensor>& inputs,
      void* stream_handle = nullptr) override;` is the interface of `AOTIModelContainerRunnerCpu`, which overrides the `AOTIModelContainerRunner`.

EikanWang commented:
> Hi @etaf, how much longer do you think you will need to land this one? Because some of the changes here are not XPU specific, and I want to make sure they can make the 2.6 branch cut (12/9).

Hi @desertfire , I have resolved all the dependencies and will land all the approved PRs of this ghstack before the PT 2.6 branch cut.

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf commented on torch/csrc/inductor/aoti_runner/model_container_runner_cuda.h:
```diff
@@ -9,7 +9,8 @@ namespace torch::inductor {
 // We provide NO BC guarantee for these APIs
 // NOLINTNEXTLINE(cppcoreguidelines-special-member-functions)
-class TORCH_API AOTIModelContainerRunnerCuda : public AOTIModelContainerRunner {
+class TORCH_CUDA_CPP_API AOTIModelContainerRunnerCuda
```
@desertfire  Her I changed `TORCH_API` to `TORCH_CUDA_CPP_API` for the following reason:
With `TORCH_API` the ciflow/trunk job cuda windows build failed with error message when linking libtorch_cuda: 
```
2024-12-06T06:09:28.8270765Z model_container_runner_cuda.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) const torch::inductor::AOTIModelContainerRunnerCuda::`vftable'" (__imp_??_7AOTIModelContainerRunnerCuda@inductor@torch@@6B@) referenced in function "public: __cdecl torch::inductor::AOTIModelContainerRunnerCuda::AOTIModelContainerRunnerCuda(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,unsigned __int64,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??0AOTIModelContainerRunnerCuda@inductor@torch@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K00@Z)
```
which shows the `vftable` is an `__declspec(dllimport)` symbol.
Then I found that `TOCH_API` is expressed as `C10_IMPORT` for lib_cuda, we should use `TORCH_CUDA_CPP_API` here.

etaf committed:
Update

[ghstack-poisoned]

etaf commented:
@pytorchbot merge

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf committed:
Update

[ghstack-poisoned]

etaf commented:
@pytorchbot merge

desertfire committed:
Update on "[AOTI] Fix #140546 and support AOTI package load for Intel GPU."

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #140686
* __->__ #140664
* #140269
* #140268
* #135320
* #135318
* #139026 


Fix #140546



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

desertfire commented:
Relanding, see https://github.com/pytorch/pytorch/pull/140269#issuecomment-2529125967

desertfire commented:
@pytorchbot merge

desertfire committed:
Update on "[AOTI] Fix #140546 and support AOTI package load for Intel GPU."

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* #140686
* __->__ #140664
* #140269
* #140268
* #135320
* #135318
* #139026 


Fix #140546



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

etaf commented:
The failed job: [xpu / linux-jammy-xpu-2025.0-py3.9 / test (default, 4, 4, linux.idc.xpu)](https://hud.pytorch.org/pr/pytorch/pytorch/140664#34163806148) ([gh](https://github.com/pytorch/pytorch/actions/runs/12245053580/job/34163806148))
inductor/test_torchinductor_opinfo.py::TestInductorOpInfoXPU::test_comprehensive_masked_cumprod_xpu_float16

is a known issue https://github.com/pytorch/pytorch/issues/141861 and has been fixed in main branch by https://github.com/pytorch/pytorch/pull/142348.

Please ignore it.

@desertfire can we land this PR after CI finished?

desertfire commented:
> The failed job: [xpu / linux-jammy-xpu-2025.0-py3.9 / test (default, 4, 4, linux.idc.xpu)](https://hud.pytorch.org/pr/pytorch/pytorch/140664#34163806148) ([gh](https://github.com/pytorch/pytorch/actions/runs/12245053580/job/34163806148)) inductor/test_torchinductor_opinfo.py::TestInductorOpInfoXPU::test_comprehensive_masked_cumprod_xpu_float16
> 
> is a known issue #141861 and has been fixed in main branch by #142348.
> 
> Please ignore it.
> 
> @desertfire can we land this PR after CI finished?

SGTM

etaf commented:
@pytorchbot merge -i
